{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from __future__ import division\n",
    "import numpy as np\n",
    "import scipy.io as sio\n",
    "from scipy import stats\n",
    "#import xgboost as xgb\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from Bio.Seq import Seq\n",
    "import pandas as pd\n",
    "import math\n",
    "import pickle\n",
    "from sklearn.metrics import roc_auc_score\n",
    "import copy\n",
    "#from libsvm.svmutil import *\n",
    "import csv\n",
    "import sys\n",
    "import re\n",
    "import scipy\n",
    "path='/Users/nhatcuong/Downloads/python_codes 2/data/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mat = scipy.io.loadmat('/Users/nhatcuong/Downloads/python_codes 2/data/classification_data_sets_new.mat')\n",
    "def rc(dna):\n",
    "    complements = str.maketrans('ACGT', 'TGCA')\n",
    "    rcseq = dna.translate(complements)[::-1]\n",
    "    return rcseq\n",
    "    \n",
    "def read_data(F_GENOME):\n",
    "    \"\"\"\n",
    "    Write your solution for Task A in this method.\n",
    "    \n",
    "    Parameters:\n",
    "        - DO NOT modify the input parameters to this function\n",
    "        - F_GENOME: A path to the file containing the FASTA-formatted genome\n",
    "    \n",
    "    Returns:\n",
    "        - A single dimension list containing discovered off-targets\n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    _TASK_A_CACHE=-1\n",
    "    if _TASK_A_CACHE > 0:\n",
    "        return _TASK_A_CACHE\n",
    "    else:\n",
    "        results = []\n",
    "        #candidateGuides=[]\n",
    "    \n",
    "        # Add your code here\n",
    "        # Opening the FASTA file and reading in the sequence\n",
    "        seq = ''\n",
    "        with open(F_GENOME,'r') as fp:\n",
    "            for line in fp:\n",
    "                if line[0] != '>':\n",
    "                    seq += line.strip()\n",
    "                    \n",
    "        #print(f\"Input sequence length: {len(seq)}\")\n",
    "    \n",
    "        opposite = rc(seq)\n",
    "\n",
    "        for pattern, strand in [\n",
    "        [r\"(?=([ATCG]{21}GG))\", '+'], \n",
    "        [r\"(?=(CC[ACGT]{21}))\", '-']]:\n",
    "            p = re.compile(pattern)\n",
    "            for m in p.finditer(seq):\n",
    "                if strand == '+':\n",
    "                    results.append(\n",
    "                        seq[m.start() : m.start() + 20] #m.group()[:-3]\n",
    "                    )\n",
    "\n",
    "                if strand == '-':\n",
    "                    results.append(\n",
    "                        rc(\n",
    "                            seq[m.start() : m.start() + 20] #m.group()[:-3]\n",
    "                        ))\n",
    "        \n",
    "        # I tried to extract guides with occurence > 1, implying that it is certain\n",
    "        # this will return a list of off targets, however, it is too expensive\n",
    "        # can=[]\n",
    "        # for guide in results:\n",
    "        #     if results.count(guide) > 1:\n",
    "        #         can.append(guide)\n",
    "        # results = results \n",
    "        #Cache the results to reduce run-time later\n",
    "\n",
    "        return results\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def countGC(input_gene, start):\n",
    "    # s = open(input_gene).read()\n",
    "    s=input_gene\n",
    "    end = len(s)\n",
    "    # 'compute the GC counts and GC contents of a sequences from (start,end)'\n",
    "    GCcounts = len(s[start:end].replace('A', '').replace('T', ''))\n",
    "    GCcontent = GCcounts / len(s)\n",
    "    return GCcounts, GCcontent\n",
    "def make_seq(input_genome):\n",
    "    seq=''\n",
    "    with open(input_genome,'r') as fp:\n",
    "        for line in fp:\n",
    "            if line[0] != '>':\n",
    "                seq += line.strip()\n",
    "    return seq\n",
    "def ncf(s, order):\n",
    "    # nucleotide composition features and position nucletide binary features\n",
    "    # s--sequence\n",
    "    # order--1,2,3: A,AA,AAA\n",
    "    t_list = [\"A\", \"G\", \"C\", \"T\"]\n",
    "    L = len(s)\n",
    "    if order == 1:\n",
    "        nc = t_list\n",
    "    elif order == 2:\n",
    "        nc = [m + n for m in [\"A\", \"G\", \"C\", \"T\"] for n in [\"A\", \"G\", \"C\", \"T\"]]\n",
    "    elif order == 3:\n",
    "        nc = [m + n + k for m in [\"A\", \"G\", \"C\", \"T\"] for n in [\"A\", \"G\", \"C\", \"T\"] for k in [\"A\", \"G\", \"C\", \"T\"]]\n",
    "    nc_f = np.zeros((1, 4 ** order))\n",
    "\n",
    "    pos_fea = np.zeros((1, 1))\n",
    "    for i in range(0, L - order + 1):\n",
    "        pos = np.zeros((1, 4 ** order))\n",
    "        for j in range(0, len(nc)):\n",
    "            if s[i:i + order] == nc[j]:\n",
    "                nc_f[0][j] = nc_f[0][j] + 1\n",
    "                pos[0][j] = 1\n",
    "                pos_fea = np.hstack((pos_fea, pos))\n",
    "\n",
    "    n = len(pos_fea[0])\n",
    "    pos_fea = pos_fea[0][1:n]\n",
    "    return nc_f, pos_fea"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "INPUT_GENOME = r\"./S_bongori.fa\" #My ending ID is 71\n",
    "INPUT_GUIDES = r\"./S_bongori.fa-27-guides.txt\" #The fifth and sixth digits of my ID are 87 ~> 27\n",
    "result = read_data(INPUT_GENOME)\n",
    "sequence = make_seq(INPUT_GENOME)\n",
    "# count_full,content_full = countGC()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "nc_f = []\n",
    "pos_fea=[]\n",
    "for seq in result[:100]:\n",
    "    nc,pos = ncf(seq,2)\n",
    "    nc_f.append(nc)\n",
    "    pos_fea.append(pos)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "count=[]\n",
    "content=[]\n",
    "for seq in result[:100]:\n",
    "    GCcount,GCcontent = countGC(seq,0)\n",
    "    count.append(GCcount)\n",
    "    content.append(GCcontent)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def binary_seqs(obs_seqs, t):\n",
    "    nts = ['A', 'C', 'G', 'T']\n",
    "    N = len(obs_seqs)\n",
    "    n = len(obs_seqs[0])\n",
    "    nc = []\n",
    "    if t == 1:\n",
    "        nc = nts\n",
    "    elif t == 2:\n",
    "        nc = [m + l for m in nts for l in nts]\n",
    "\n",
    "    alpStat = np.zeros((N, 1 + n - t))\n",
    "    for i in range(0, N):\n",
    "        seq = obs_seqs[i]\n",
    "        for j in range(0, 1 + n - t):\n",
    "            mer = seq[j:j + t]\n",
    "            ind = nc.index(mer)\n",
    "            alpStat[i, j] = ind\n",
    "\n",
    "    return alpStat\n",
    "alpStat=[]\n",
    "for seq in result[:100]:\n",
    "    a = binary_seqs(seq,2)\n",
    "    alpStat.append(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sequences(fname):\n",
    "    df = pd.read_excel(fname)\n",
    "    targets = df['seq'].apply(lambda x: x.split(\",\")[0][:20]).unique()\n",
    "    # targets = map(lambda x: (\"NONE\", x.upper(), 'AGG'), targets)\n",
    "    return targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "fname=path+'guide_text.xlsx'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "target = get_sequences(fname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GAACACAAAGCATAGACTGC', 'GACCCCCTCCACCCCGCCTC',\n",
       "       'GAGTCCGAGCAGAAGAAGAA', 'GGAATCCCTTCTGCAGCACC',\n",
       "       'GGCACTGCGGCTGGAGGTGG', 'GGCCCAGACTGAGCACGTGA',\n",
       "       'GGGAAAGACCCAGCATCCGT', 'GGGTGGGGGGAGTTTGCTCC',\n",
       "       'GGTGAGTGAGTGTGTGCGTG'], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/nhatcuong/opt/anaconda3/lib/python3.8/site-packages/openpyxl/worksheet/_reader.py:312: UserWarning: Unknown extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "v1 =pd.read_excel(path+\"V1_data.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Target</th>\n",
       "      <th>30mer</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>CutSite</th>\n",
       "      <th>Annotation</th>\n",
       "      <th>MOLM13 CD15</th>\n",
       "      <th>MOLM13 CD33</th>\n",
       "      <th>NB4 CD33</th>\n",
       "      <th>TF1 CD33</th>\n",
       "      <th>NB4 CD13</th>\n",
       "      <th>TF1 CD13</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACGTACAGGCCCCTGTCATT</td>\n",
       "      <td>CD13</td>\n",
       "      <td>AAAAACGTACAGGCCCCTGTCATTGGGGGTGAGG</td>\n",
       "      <td>antisense</td>\n",
       "      <td>ENST00000300060</td>\n",
       "      <td>8594</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.367491</td>\n",
       "      <td>-1.014438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGTACAGGCCCCTGTCATTG</td>\n",
       "      <td>CD13</td>\n",
       "      <td>AAAACGTACAGGCCCCTGTCATTGGGGGTGAGGT</td>\n",
       "      <td>antisense</td>\n",
       "      <td>ENST00000300060</td>\n",
       "      <td>8593</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.291343</td>\n",
       "      <td>0.785962</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GTACAGGCCCCTGTCATTGG</td>\n",
       "      <td>CD13</td>\n",
       "      <td>AAACGTACAGGCCCCTGTCATTGGGGGTGAGGTA</td>\n",
       "      <td>antisense</td>\n",
       "      <td>ENST00000300060</td>\n",
       "      <td>8592</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.493360</td>\n",
       "      <td>2.035237</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CGCTGAAACCCGATTCCTAC</td>\n",
       "      <td>CD13</td>\n",
       "      <td>AACACGCTGAAACCCGATTCCTACCGGGTGACGC</td>\n",
       "      <td>sense</td>\n",
       "      <td>ENST00000300060</td>\n",
       "      <td>8558</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.789269</td>\n",
       "      <td>-0.997515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCAATGTGACGGGCTATTAC</td>\n",
       "      <td>CD13</td>\n",
       "      <td>AACCTCAATGTGACGGGCTATTACCGGGTGAACT</td>\n",
       "      <td>sense</td>\n",
       "      <td>ENST00000300060</td>\n",
       "      <td>15398</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.683439</td>\n",
       "      <td>-0.418296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>877</th>\n",
       "      <td>CCTCACTAGACTTGACCCAC</td>\n",
       "      <td>CD33</td>\n",
       "      <td>TTCTCCTCACTAGACTTGACCCACAGGCCCAAAA</td>\n",
       "      <td>sense</td>\n",
       "      <td>ENST00000262262</td>\n",
       "      <td>739</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.681098</td>\n",
       "      <td>2.597234</td>\n",
       "      <td>2.843958</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878</th>\n",
       "      <td>TGTCGTAGTAGGGTATGGGA</td>\n",
       "      <td>CD33</td>\n",
       "      <td>TTCTTGTCGTAGTAGGGTATGGGATGGAAGAAAG</td>\n",
       "      <td>antisense</td>\n",
       "      <td>ENST00000262262</td>\n",
       "      <td>246</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.970595</td>\n",
       "      <td>2.813329</td>\n",
       "      <td>3.004306</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>879</th>\n",
       "      <td>TCTAGCTTGTTTGTGGCCAC</td>\n",
       "      <td>CD33</td>\n",
       "      <td>TTGATCTAGCTTGTTTGTGGCCACTGGAGAGTCC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>ENST00000262262</td>\n",
       "      <td>328</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.329461</td>\n",
       "      <td>2.245616</td>\n",
       "      <td>2.635855</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>880</th>\n",
       "      <td>TGCCGTGGTCCTGGGGCCGT</td>\n",
       "      <td>CD33</td>\n",
       "      <td>TTGGTGCCGTGGTCCTGGGGCCGTGGGGTGATTA</td>\n",
       "      <td>antisense</td>\n",
       "      <td>ENST00000262262</td>\n",
       "      <td>918</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.808257</td>\n",
       "      <td>2.370927</td>\n",
       "      <td>2.828231</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>881</th>\n",
       "      <td>TCCTACTGGGATCCCCAAGG</td>\n",
       "      <td>CD33</td>\n",
       "      <td>TTGTTCCTACTGGGATCCCCAAGGAGGCGGAATC</td>\n",
       "      <td>antisense</td>\n",
       "      <td>ENST00000262262</td>\n",
       "      <td>387</td>\n",
       "      <td>CDS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.564535</td>\n",
       "      <td>2.382079</td>\n",
       "      <td>2.926046</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>882 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 Sequence Target                               30mer  \\\n",
       "0    ACGTACAGGCCCCTGTCATT   CD13  AAAAACGTACAGGCCCCTGTCATTGGGGGTGAGG   \n",
       "1    CGTACAGGCCCCTGTCATTG   CD13  AAAACGTACAGGCCCCTGTCATTGGGGGTGAGGT   \n",
       "2    GTACAGGCCCCTGTCATTGG   CD13  AAACGTACAGGCCCCTGTCATTGGGGGTGAGGTA   \n",
       "3    CGCTGAAACCCGATTCCTAC   CD13  AACACGCTGAAACCCGATTCCTACCGGGTGACGC   \n",
       "4    TCAATGTGACGGGCTATTAC   CD13  AACCTCAATGTGACGGGCTATTACCGGGTGAACT   \n",
       "..                    ...    ...                                 ...   \n",
       "877  CCTCACTAGACTTGACCCAC   CD33  TTCTCCTCACTAGACTTGACCCACAGGCCCAAAA   \n",
       "878  TGTCGTAGTAGGGTATGGGA   CD33  TTCTTGTCGTAGTAGGGTATGGGATGGAAGAAAG   \n",
       "879  TCTAGCTTGTTTGTGGCCAC   CD33  TTGATCTAGCTTGTTTGTGGCCACTGGAGAGTCC   \n",
       "880  TGCCGTGGTCCTGGGGCCGT   CD33  TTGGTGCCGTGGTCCTGGGGCCGTGGGGTGATTA   \n",
       "881  TCCTACTGGGATCCCCAAGG   CD33  TTGTTCCTACTGGGATCCCCAAGGAGGCGGAATC   \n",
       "\n",
       "        Strand       Transcript  CutSite Annotation  MOLM13 CD15  MOLM13 CD33  \\\n",
       "0    antisense  ENST00000300060     8594        CDS          NaN          NaN   \n",
       "1    antisense  ENST00000300060     8593        CDS          NaN          NaN   \n",
       "2    antisense  ENST00000300060     8592        CDS          NaN          NaN   \n",
       "3        sense  ENST00000300060     8558        CDS          NaN          NaN   \n",
       "4        sense  ENST00000300060    15398        CDS          NaN          NaN   \n",
       "..         ...              ...      ...        ...          ...          ...   \n",
       "877      sense  ENST00000262262      739        CDS          NaN     2.681098   \n",
       "878  antisense  ENST00000262262      246        CDS          NaN     2.970595   \n",
       "879  antisense  ENST00000262262      328        CDS          NaN     2.329461   \n",
       "880  antisense  ENST00000262262      918        CDS          NaN     1.808257   \n",
       "881  antisense  ENST00000262262      387        CDS          NaN     2.564535   \n",
       "\n",
       "     NB4 CD33  TF1 CD33  NB4 CD13  TF1 CD13  \n",
       "0         NaN       NaN  0.367491 -1.014438  \n",
       "1         NaN       NaN  1.291343  0.785962  \n",
       "2         NaN       NaN  1.493360  2.035237  \n",
       "3         NaN       NaN -0.789269 -0.997515  \n",
       "4         NaN       NaN  0.683439 -0.418296  \n",
       "..        ...       ...       ...       ...  \n",
       "877  2.597234  2.843958       NaN       NaN  \n",
       "878  2.813329  3.004306       NaN       NaN  \n",
       "879  2.245616  2.635855       NaN       NaN  \n",
       "880  2.370927  2.828231       NaN       NaN  \n",
       "881  2.382079  2.926046       NaN       NaN  \n",
       "\n",
       "[882 rows x 13 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "v1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "al=binary_seqs(result[1],1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "def TranEmis(alpStat, t, au=5, ad=5):\n",
    "    N = len(alpStat[:, 0])\n",
    "    n = len(alpStat[0, :])\n",
    "    b = np.zeros((1, n + 2))\n",
    "    c = np.zeros((n + 1, 1))\n",
    "    d = np.eye(n + 1)\n",
    "    Tr = np.hstack((c, d))\n",
    "    Tr = np.vstack((Tr, b))\n",
    "    tr = Tr.copy()\n",
    "    tr[tr == 0] = float('-inf')\n",
    "    tr[tr == 1] = 0\n",
    "\n",
    "    Et = np.zeros((n + 2, 4 ** t))\n",
    "    et = Et.copy()\n",
    "    et[et == 0] = float('-inf')\n",
    "    for i in range(1, n + 1):\n",
    "        for j in range(0, 4 ** t):\n",
    "            count = np.where(alpStat[:, i - 1] == j)\n",
    "            p = (len(count[0]) + au) / (N + ad)\n",
    "            if p == 0:\n",
    "                et[i, j] = -0.01\n",
    "            else:\n",
    "                et[i, j] = math.log(p)\n",
    "\n",
    "    return tr, et"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([[-inf,   0., -inf],\n",
       "        [-inf, -inf,   0.],\n",
       "        [-inf, -inf, -inf]]),\n",
       " array([[       -inf,        -inf,        -inf,        -inf],\n",
       "        [-0.98082925, -0.69314718, -0.98082925, -0.87546874],\n",
       "        [       -inf,        -inf,        -inf,        -inf]]))"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TranEmis(al,1,au=5,ad=4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "70617bfe65eff824b4c75cd1b33791dc69622f376b1e77c05a1d0d94907355cf"
  },
  "kernelspec": {
   "display_name": "Python 3.8.8 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
